delta-lake/delta-23x/src/main/scala/org/apache/spark/sql/delta/rapids/delta23x/GpuUpdateCommand.scala

```diff
@@ -0,0 +1,275 @@
+    "numTouchedRows" -> createMetric(sc, "number of rows touched (copied + updated)")
+  )
+
+  final override def run(sparkSession: SparkSession): Seq[Row] = {
+    recordDeltaOperation(tahoeFileIndex.deltaLog, "delta.dml.update") {
+      val deltaLog = tahoeFileIndex.deltaLog
+      gpuDeltaLog.withNewTransaction { txn =>
+        DeltaLog.assertRemovable(txn.snapshot)
+        if (hasBeenExecuted(txn, sparkSession)) {
+          sendDriverMetrics(sparkSession, metrics)
+          return Seq.empty
+        }
+        performUpdate(sparkSession, deltaLog, txn)
+      }
+      // Re-cache all cached plans(including this relation itself, if it's cached) that refer to
+      // this data source relation.
+      sparkSession.sharedState.cacheManager.recacheByPlan(sparkSession, target)
+    }
+    Seq(Row(metrics("numUpdatedRows").value))
+  }
+
+  private def performUpdate(
+      sparkSession: SparkSession, deltaLog: DeltaLog, txn: OptimisticTransaction): Unit = {
+    import org.apache.spark.sql.delta.implicits._
+
+    var numTouchedFiles: Long = 0
+    var numRewrittenFiles: Long = 0
+    var numAddedBytes: Long = 0
+    var numRemovedBytes: Long = 0
+    var numAddedChangeFiles: Long = 0
+    var changeFileBytes: Long = 0
+    var scanTimeMs: Long = 0
+    var rewriteTimeMs: Long = 0
+
+    val startTime = System.nanoTime()
+    val numFilesTotal = txn.snapshot.numOfFiles
+
+    val updateCondition = condition.getOrElse(Literal.TrueLiteral)
+    val (metadataPredicates, dataPredicates) =
+      DeltaTableUtils.splitMetadataAndDataPredicates(
+        updateCondition, txn.metadata.partitionColumns, sparkSession)
+    val candidateFiles = txn.filterFiles(metadataPredicates ++ dataPredicates)
+    val nameToAddFile = generateCandidateFileMap(deltaLog.dataPath, candidateFiles)
+
+    scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
+
+    val filesToRewrite: Seq[AddFile] = if (candidateFiles.isEmpty) {
+      // Case 1: Do nothing if no row qualifies the partition predicates
+      // that are part of Update condition
+      Nil
+    } else if (dataPredicates.isEmpty) {
+      // Case 2: Update all the rows from the files that are in the specified partitions
+      // when the data filter is empty
+      candidateFiles
+    } else {
+      // Case 3: Find all the affected files using the user-specified condition
+      val fileIndex = new TahoeBatchFileIndex(
+        sparkSession, "update", candidateFiles, deltaLog, tahoeFileIndex.path, txn.snapshot)
+      // Keep everything from the resolved target except a new TahoeFileIndex
+      // that only involves the affected files instead of all files.
+      val newTarget = DeltaTableUtils.replaceFileIndex(target, fileIndex)
+      val data = Dataset.ofRows(sparkSession, newTarget)
+      val updatedRowCount = metrics("numUpdatedRows")
+      val updatedRowUdf = DeltaUDF.boolean {
+        new GpuDeltaMetricUpdateUDF(updatedRowCount)
+      }.asNondeterministic()
```

Comment:
Do we want this as nondeterministic? We have it this way in 2.4 too. Just curious if it might be a bug or not.

Reply:
Having this be non-deterministic matches the CPU Delta Lake behavior.  It's OK in this case, as it's applied as part of a filter (see just a few lines below).  We're only marking the other UDFs as deterministic because they appear in a non-project, non-filter scenario due to their appearance as sub-expressions of the new RapidsProcessDeltaMergeJoin node.